### ISSUE TITLE
Doubts about DL3059

### ISSUE THREAD

**Issue by dominikbraun** [hadolint/hadolint#715]
So I have some doubts regarding [DL3059](https://github.com/hadolint/hadolint/wiki/DL3059).

My Dockerfile looks like this:

```Dockerfile
RUN apk add --no-cache \
    curl \
    tar


RUN curl -LO https://github.com/dominikbraun/timetrace/releases/download/${VERSION}/timetrace-linux-amd64.tar.gz && \
    tar -xvf timetrace-linux-amd64.tar.gz -C /bin && \
    rm -f timetrace-linux-amd64.tar.gz
```

**DL3059 suggests:**

```
One of the first points in the Docker best practices is that one should merge
multiple RUN instructions into one, where possible to reduce the layer count.
```

I don't think that this is a universal truth.

1. It is true that merging multiple `RUN` instructions reduces the _layer count_, but that doesn't mean that it reduces _image size_. The files produced by two or more commands won't be any different when these commands are executed in a single step (I wrote an [in-depth article](https://dominikbraun.io/blog/docker/docker-images-and-their-layers-explained/) on this topic).
2. Merging multiple `RUN` instructions merely is a mechanism to prevent caching issues, for example when installing packages (see the [`RUN` Best Practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#run)).
3. Most importantly: The abovementioned mechanism is referred to as cache busting since it forces Docker to re-run the instructions. Merging `RUN` instructions as a general "Best Practice" probably decreases cache hits drastically. This is also mentioned in the reference for DL3059.

To sum it up: If I'm not missing something, DL3059 won't lead to smaller images but just to longer builds since the cache can be used less often. I doubt that this is a good general advice.

**Comment by m-ildefons** [comment]
Hi @dominikbraun ,

that is true. DL3059 is a controversial rule and disabling it is a valid choice. However I'd like to point out that it does prevent some problems that can easily trip developers and especially Docker-newbies up. For example:

```Dockerfile
RUN apt-get update
RUN apt-get install -y --no-install-recommends curl tar
RUN curl -LO https://foo.bar/download/this.tar
RUN tar -xf this.tar
RUN rm -f this.tar
```
There are multiple problems here:
- Not running the `curl` and the `rm` commands in the same layer causes that middle layer to be unnecessarily large, which would also cause the total image size to be unnecessarily large (correct me if I'm wrong).
- caching the `apt-get update` in a different layer from the `apt-get install` may lead to problems where the package list remains outdated and thus dependencies are 'missing'. I know from first hand experience that SUSEs `zypper` is prone to that problem and that it is very pronounced in rolling release skews of distributions like `opensuse/tubleweed`. This problem would actually cause a build to fail entirely.
- Running each command in a separate shell may also cause problems e.g. with the environment variables

I know that this is an extreme example, but I have seen experienced developers make these mistakes simply because they are unfamiliar with the intricacies of Docker. Usually a longer build time is tolerable if the end result is good.
That being said, I admit that this rule is probably too general. Maybe it should not trigger if one of the `RUN` instructions already chains commands together with `&&` or `;`? What do you think, would that work?

Best regards, Moritz

PS: There is a fourth edge case where DL3059 is unhelpful:
In case your Docker registry has an upload size limit for the layers, say 2GB, one can run into problems where installing too many huge packages (e.g. LLVM) can put you over that limit, but breaking it up into several small layers is fine.

**Comment by dominikbraun** [comment]
@m-ildefons Thanks for your response!

All your points are perfectly valid. Your `apt-get update` example probably is the most typical use case for merging `RUN` instructions. If the user forgets about this or doesn't know how layer caching works, DL3059 is indeed helpful.

> That being said, I admit that this rule is probably too general. Maybe it should not trigger if one of the RUN instructions already chains commands together with && or ;? What do you think, would that work?

Yes, I think the actual problem is that hadolint complains about separate consecutive `RUN` instructions in general. Your solution with a check for `RUN` instructions concatenated with `&&` sounds good, because if I'm using such instructions, I'm obviously aware of the merging mechanism.

If I had the choice, I probably would either take that approach or reduce the level of DL3059, or both - but that's really up to you. I just opened this issue because I was wondering that hadolint complains about separated `RUN` instructions even though I exactly knew what I was doing.

**Comment by mjambon** [comment]
I edited the wiki page (examples + rationale) to explain when space savings occur: https://github.com/hadolint/hadolint/wiki/DL3059/_compare/d14175ac837fdb52479c3b76fa395259869874c7...4d6ffebadfc4a63724fa9b10e85552ca256a01d0

**Comment by jjmaestro** [comment]
@m-ildefons I understand your explanations and I find them quite reasonable. I'm personally overriding `DL3059` to be `style` and IMHO I think that makes more sense as it will be shown to the user when running `hadolint` but e.g. it won't break linters. What do you think?
